js/plugin/POIMarkers.js

```diff
@@ -0,0 +1,138 @@
+
+    options: {
+        routing: null,
+        shortcut: {
+            draw: {
+                enable: 80, // char code for 'p'
+                disable: 27 // char code for 'ESC'
+            }
+        }
+    },
+
+    onAdd: function(map) {
+        var self = this;
+
+        this.map = map;
+        this.markersLayer = L.layerGroup([]).addTo(map);
+
+        this.drawButton = L.easyButton({
+            states: [
+                {
+                    stateName: 'activate-poi',
+                    icon: 'fa-hand-o-right',
+                    onClick: function() {
+                        self.draw(true);
+                    },
+                    title: i18next.t('map.draw-poi-start')
+                },
+                {
+                    stateName: 'deactivate-poi',
+                    icon: 'fa-hand-o-right active',
+                    onClick: function() {
+                        self.draw(false);
+                    },
+                    title: i18next.t('map.draw-poi-stop')
+                }
+            ]
+        }).addTo(map);
+
+        map.on('routing:draw-start', function() {
+            self.draw(false);
+        });
+
+        var container = new L.DomUtil.create('div');
+        // keys not working when map container does not have focus, use document instead
+        L.DomEvent.removeListener(container, 'keyup', this._keyupListener);
+        L.DomEvent.addListener(document, 'keyup', this._keyupListener, this);
+
+        return container;
+    },
+
+    draw: function(enable) {
+        this.drawButton.state(enable ? 'deactivate-poi' : 'activate-poi');
+        if (enable) {
+            this.options.routing.draw(false);
+            this.map.on('click', this.onMapClick, this);
+            L.DomUtil.addClass(this.map.getContainer(), 'pois-draw-enabled');
+        } else {
+            this.map.off('click', this.onMapClick, this);
+            L.DomUtil.removeClass(this.map.getContainer(), 'pois-draw-enabled');
+        }
+    },
+
+    _keyupListener: function(e) {
+        // Suppress shortcut handling when a text input field is focussed
+        if (document.activeElement.type == 'text' || document.activeElement.type == 'textarea') {
+            return;
+        }
+        if (e.keyCode === this.options.shortcut.draw.disable) {
+            this.draw(false);
+        } else if (e.keyCode === this.options.shortcut.draw.enable) {
+            this.draw(true);
+        }
+    },
+
+    onMapClick: function(e) {
+        var self = this;
+        bootbox.prompt({
+            title: i18next.t('map.enter-poi-name'),
+            callback: function(result) {
+                if (result !== null) {
+                    self.addMarker(e.latlng, result);
+                }
+            }
+        });
+    },
+
+    addMarker: function(latlng, name) {
+        var icon = L.AwesomeMarkers.icon({
+            icon: 'star',
+            markerColor: 'cadetblue'
+        });
+        var content = name + '<br>';
```

Comment:
The user input `name` should be sanitized/encoded here to prevent [Cross-site Scripting (XSS)](https://www.owasp.org/index.php/Cross-site_Scripting_(XSS)) (DOM based).

For example this URL injects a JavaScript alert that is executed when the popup opens:
http://localhost:3000/#map=6/47.636/7.086/standard&pois=7.58045,47.576526,%3Cimg%20src%3D%22xyz%22%20onerror%3D%22alert('script%20executed')%22%3E

The most simple fix usually is to use `Node.textContent` instead of `innerHTML`, but in this case it is used by Leaflet internally in [DivOverlay.js](https://github.com/Leaflet/Leaflet/blob/56cdbf69a91465e1e1db047e17cc388e53a74652/src/layer/DivOverlay.js#L201). We could also use a lib like [DOMPurify](https://github.com/cure53/DOMPurify).

Reply:
Regarding [this article](https://gomakethings.com/preventing-cross-site-scripting-attacks-when-using-innerhtml-in-vanilla-javascript/), I added the provided sanitizer function: 

```js
var sanitizeHTML = function (str) {
            var temp = document.createElement('div');
            temp.textContent = str;
            return temp.innerHTML;
        };
```
WDYT?

Reply:
Will probably work for our use case, but not when used for attribute values, according to that old article:
[You are probably misusing DOM text methods](https://web.archive.org/web/20121208091505/http://benv.ca/2012/10/4/you-are-probably-misusing-DOM-text-methods/)

So maybe rename to `sanitizeHTMLContent` and add a comment to not use for attributes.

Reply:
Since the method is internal to POI markers, I believe we won't use it for something else but indeed I added a comment about that. Thanks!